dvc/repo/__init__.py

```diff
@@ -454,14 +455,23 @@ def is_dvc_internal(self, path):
     @contextmanager
     def open(self, path, remote=None, mode="r", encoding=None):
         """Opens a specified resource as a file descriptor"""
+        if not remote and self.tree.exists(path):
+            with self.tree.open(path, mode, encoding) as fd:
+                yield fd
+            return
+
         try:
-            with self._open(path, remote, mode, encoding) as fd:
+            with self._open_out(path, remote, mode, encoding) as fd:
                 yield fd
-        except FileNotFoundError:
-            raise OutputFileMissingError(relpath(path, self.root_dir))
+        except FileNotFoundError as e:
+            raise OutputFileMissingError(relpath(path, self.root_dir), cause=e)
 
-    def _open(self, path, remote=None, mode="r", encoding=None):
+    def _open_out(self, path, remote=None, mode="r", encoding=None):
         out, = self.find_outs_by_path(path)
+
+        if not out.use_cache and not out.exists:
```

Comment:
This condition is wrong - if out is not using cache, but is missing we continue to try open it from cache, which doesn't make sense.

I would say we should check if out is cached before `._open_out()` then, rename `._open_out()` to `._open_cached()` and pass `out` there instead of path.